add backoff retry on service check

Omarabdul3ziz opened the pull request:
### Description
add backoff retry on service check to pass time if the network is not setup yet


### Related Issues
- https://github.com/threefoldtech/zos/issues/2255

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

Omarabdul3ziz committed:
add backoff retry on service check to pass time if the network is not setup yet

muhamadazmy reviewed (changes requested):
This is not the right approach to fix this. Your solution will NOT show any error to the user until the issue is resolved.

The correct logic is that if we have errors, we need to show them to the user (via the PushErrors) then retry again

So while a backoff strategy is a good idea, this solution does not satisfy the requirements. You need to take another look

Hint: the retry has to be done on a higher layer

Omarabdul3ziz committed:
- directly push errors to zui after running each check
- move the retrying to an upper level to retry the whole check with pushing to zui

Omarabdul3ziz committed:
cleaner err check

muhamadazmy commented on pkg/perf/healthcheck/healthcheck.go:
```diff
@@ -60,21 +62,35 @@ func (h *healthcheckTask) Run(ctx context.Context) (interface{}, error) {
-			continue
+		op := func() error {
+			errors := check(ctx)
+			errs[label] = errorsToStrings(errors)
```
why the map? why do you need to use `errs` here this way?

Omarabdul3ziz replied:
errs are returned at the end as the result of the health check task, this is what is stored on redis and responded to the rmb call asking for the health check

muhamadazmy commented on pkg/perf/healthcheck/healthcheck.go:
```diff
@@ -60,21 +62,35 @@ func (h *healthcheckTask) Run(ctx context.Context) (interface{}, error) {
-		err := zui.PushErrors(ctx, label, errs[label])
-		if err != nil {
+		if err := backoff.RetryNotify(op, bo, notify); err != nil {
 			return nil, err
```
I don't believe you should return here, since there are other checks you need to do.

muhamadazmy commented on pkg/perf/healthcheck/healthcheck.go:
```diff
@@ -60,21 +62,35 @@ func (h *healthcheckTask) Run(ctx context.Context) (interface{}, error) {
+		bo := backoff.NewExponentialBackOff()
+		bo.InitialInterval = 30 * time.Second
+		bo.MaxInterval = 30 * time.Second
+		bo.MaxElapsedTime = 10 * time.Minute
```
hmm, i am wondering if this is a good number. Note that this will block other checks from running. I am wondering if we should start all checks in parallel so they don't block each other (since they are not related) then each check can has it's own 10 min retry loop

Omarabdul3ziz committed:
run each health check in separate routine
